File: Calabrigo/Informatica/java/ProblemiDaConsegnare/Problema10/src/StringSList.java

public class StringSList {
	boolean empty;
	StringSList rest;
	String first;
	
	public StringSList() {
		empty = true;
		rest = null;
	}
	
	public StringSList(String first,StringSList rest) {
		this.first = first;
		this.rest = rest;
		this.empty = false;
	}
	
	public String Car() {
		return first;
	}
	
	public StringSList Cdr() {
		return rest;
	}
	
	public StringSList Cons(String str) {
		return new StringSList(str ,this);
	}
	
	public int Length() {
		StringSList q = this;
		int counter = 0;
		while(q.rest != null) {
			q = q.Cdr();
			counter++;
		}
		return counter + 1;
	}
	
	public String StringRef(int index) {
		StringSList q = this;
		int counter = 0;
		while(counter != index) {
			q = q.Cdr();
			counter++;
		}
		return q.first;
	}
	
	public String toString() {
		String v = "";
		v = v + "(" + this.Car();
		StringSList q = this.rest;
		while(q != null) {
			v = v + ", " + q.Car();
			q = q.Cdr();
		}
		return v + ")";
	}
	
	public StringSList Append(StringSList list) {
		StringSList q = this;
		while(list != null) {
			q = q.Cons(list.Car());
			list = list.Cdr();
		}
		return q;
	}
	
	public StringSList Reverse() {
		StringSList q = this;
		StringSList r = new StringSList();
		while(q != null) {
			r = r.Cons(q.Car());
			q = q.Cdr();
		}
		return r;
	}

}

File: Rosu/programmazione/laboratorio/java/problemi/problema10/src/ro/razvan/uniud/stringSList/StringSList.kt
package ro.razvan.uniud.stringSList

@Suppress("EqualsOrHashCode")
class StringSList {

    private val empty: Boolean
    private val first: String
    private val rest: StringSList?

    companion object {
        val NULL_STRING_S_LIST = StringSList()
    }

    public constructor() {
        empty = true
        first = ""
        rest = null
    }

    constructor(first: String, sl: StringSList = NULL_STRING_S_LIST) {
        rest = sl
        this.first = first
        empty = false
    }

    fun isNull(): Boolean {
        return empty
    }

    val car: String
        get() = first

    val cdr: StringSList
        get() = rest ?: NULL_STRING_S_LIST

    fun cons(newElement: String): StringSList {
        return if (empty) {
            StringSList(newElement)
        } else {
            return StringSList(newElement, this)
        }
    }

    val length: Int
        get() = if (empty) {
            0
        } else {
            1 + cdr.length
        }

    fun listRef(k: Int): String {
        return if (k == 0) {
            first
        } else {
            cdr.listRef(k - 1)
        }
    }

    override fun equals(other: Any?): Boolean {
        return when {
            other == null || other !is StringSList -> false
            other.isNull() || empty -> empty && other.isNull()
            other.car != car -> false
            else -> other.cdr == cdr
        }
    }

    fun append(sl: StringSList): StringSList {
        return if (empty) {
            sl
        } else {
            cdr.append(sl).cons(first)
        }
    }

    fun reverse(): StringSList {
        return reverseRec(StringSList())
    }

    private fun reverseRec(sl: StringSList): StringSList {
        return if (empty) {
            sl
        } else {
            cdr.reverseRec(sl.cons(first))
        }
    }

    override fun toString(): String {

        return when {
            empty -> "()"
            cdr.empty -> "($first)"
            else -> {
                val builder = StringBuilder("($first")
                var p = cdr

                while (!p.isNull()) {
                    builder.append(", ${p.first}")
                    p = p.cdr
                }

                builder.append(")").toString()
            }
        }

    }

}